File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P11.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.AbstractMap.SimpleEntry;
import java.util.List;

import static java.util.stream.Collectors.toList;

/**
 * <b>(*) Modified run-length encoding</b>
 * <pre>
 *      encode_modified(Arrays.asList("a", "a", "a", "a", "b", "c", "c", "a", "a", "d", "e", "e", "e", "e"))
 * </pre>
 */
public class P11 {

    public static <T> List<Object> encode_modified(List<T> list) {
        return P09.pack(list).stream().map(l -> {
            if (l.size() == 1) {
                return l.get(0);
            }
            return new SimpleEntry<>(l.size(), l.get(0));
        }).collect(toList());
    }
}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P11.scala
package com.shekhargulati.ninetynine_problems._01_lists

/**
  * <b>(*) Modified run-length encoding</b>
  * <p>
  * Modify the result of problem 1.10 in such a way that if an element has no duplicates it is simply copied into the result list.
  * Only elements with duplicates are transferred as [N,E] terms.
  * </p>
  * {{{
  *   encode_modified(List("a", "a", "a", "a", "b", "c", "c", "a", "a", "d", "e", "e", "e", "e"))
  * }}}
  *
  */
object P11 {

  def encode_modified[T](list: List[T]): List[Any] = {
    P09.pack_span(list).map(ll => ll match {
      case x :: Nil => x
      case x :: xs => (ll.length, x)
      case Nil => Nil
    })
  }

}
